aten/src/ATen/core/ivalue_inl.h

```diff
@@ -159,40 +160,40 @@ struct CAFFE2_API ConstantString final : c10::intrusive_ptr_target {
 template <typename Elem>
 struct CAFFE2_API List : c10::intrusive_ptr_target {
  private:
-  std::vector<Elem> elements_;
+  c10::ListPtr<Elem> elements_;
```

Comment:
something doesn't add up - List itself is a intrusive_ptr_target and it stores another intrusive_ptr_target inside for elements_.impl_. Something is wrong here.

I think it makes sense to store ListImpl in IValue and instead of the wrapper here return ListPtr from IValue (it'd require a change of API - need to think whether it's ok). The benefit is that we wouldn't need to put annoying .elements() everywhere (and we can add fake .elements() returning itself for BC).

The same actually applies to Dicts as well - I missed it on that PR

Reply:
Yes, this is currently a double indirection which isn't great. I have it on my todo list to investigate if we can fix that.